components/feature/serviceworker/src/main/java/mozilla/components/feature/serviceworker/ServiceWorkerSupport.kt

```diff
@@ -0,0 +1,50 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.serviceworker
+
+import androidx.annotation.VisibleForTesting
+import mozilla.components.browser.state.state.SessionState
+import mozilla.components.concept.engine.Engine
+import mozilla.components.concept.engine.EngineSession
+import mozilla.components.concept.engine.EngineSession.LoadUrlFlags
+import mozilla.components.concept.engine.serviceworker.ServiceWorkerDelegate
+import mozilla.components.feature.tabs.TabsUseCases.AddNewTabUseCase
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * Add support for all [Engine] registered service workers' events and callbacks.
+ * Call [install] to actually start observing them.
+ */
+object ServiceWorkerSupport {
+    @VisibleForTesting internal lateinit var addTabUseCase: AddNewTabUseCase
+    private val logger = Logger("ServiceWorkerSupport")
+
+    /**
+     * Start observing service workers' events and callbacks.
+     *
+     * @param engine [Engine] for which to observe service workers events and callbacks.
+     * @param addTabUseCase [AddNewTabUseCase] delegate for opening new tabs when requested by service workers.
+     */
+    fun install(
+        engine: Engine,
+        addTabUseCase: AddNewTabUseCase
+    ) {
+        try {
+            engine.registerServiceWorkerDelegate(object : ServiceWorkerDelegate {
+                override fun addNewTab(engineSession: EngineSession): Boolean {
+                    addTabUseCase(
+                        flags = LoadUrlFlags.external(),
+                        engineSession = engineSession,
+                        source = SessionState.Source.Internal.None
```

Comment:
Otherwise from `AddNewTabUseCase` [source would be `Internal.NewTab`](https://github.com/mozilla-mobile/android-components/blob/ac74a577b68be269a7d86ed753d176d6d1c33a45/components/feature/tabs/src/main/java/mozilla/components/feature/tabs/TabsUseCases.kt#L154) though for `Internal` the [default should be `None`](https://github.com/mozilla-mobile/android-components/blob/ac74a577b68be269a7d86ed753d176d6d1c33a45/components/browser/state/src/main/java/mozilla/components/browser/state/state/SessionState.kt#L130-L133)

Seems like we need [a new `External` source](https://github.com/mozilla-mobile/android-components/blob/ac74a577b68be269a7d86ed753d176d6d1c33a45/components/browser/state/src/main/java/mozilla/components/browser/state/state/SessionState.kt#L86-L109) for this requests coming from the Engine.

Reply:
Yes, this could be a follow-up. We're wouldn't be using it right now anyway, and `source` needs more clean-up anyway.